packages/firestore/src/local/memory_document_overlay_cache.ts

```diff
@@ -0,0 +1,183 @@
+      this.overlayByBatchId.get(existing.largestBatchId).delete(mutation.key);
+    }
+
+    this.overlays = this.overlays.insert(
+      mutation.key,
+      new Overlay(largestBatchId, mutation)
+    );
+
+    // Create the association of this overlay to the given largestBatchId.
+    if (this.overlayByBatchId.get(largestBatchId) === undefined) {
+      this.overlayByBatchId.set(largestBatchId, new Set<DocumentKey>());
+    }
+    // @ts-ignore: we know the result of `get` is not undefined.
+    this.overlayByBatchId.get(largestBatchId).add(mutation.key);
+  }
+
+  saveOverlays(
+    transaction: PersistenceTransaction,
+    largestBatchId: number,
+    overlays: Map<DocumentKey, Mutation>
+  ): PersistencePromise<void> {
+    overlays.forEach((mutation, key) => {
+      this.saveOverlay(transaction, largestBatchId, mutation);
+    });
+    return PersistencePromise.resolve();
+  }
+
+  removeOverlaysForBatchId(
+    transaction: PersistenceTransaction,
+    batchId: number
+  ): PersistencePromise<void> {
+    if (this.overlayByBatchId.has(batchId)) {
+      const keys = this.overlayByBatchId.get(batchId);
+      this.overlayByBatchId.delete(batchId);
+      // @ts-ignore: we know keys is not undefined
+      for (const key of keys) {
+        this.overlays = this.overlays.remove(key);
+      }
+    }
+    return PersistencePromise.resolve();
+  }
+
+  getOverlaysForCollection(
+    transaction: PersistenceTransaction,
+    collection: ResourcePath,
+    sinceBatchId: number
+  ): PersistencePromise<Map<DocumentKey, Overlay>> {
+    const result: Map<DocumentKey, Overlay> = new Map<DocumentKey, Overlay>();
+
+    const immediateChildrenPathLength = collection.length + 1;
+    const prefix = new DocumentKey(collection.child(''));
+    const iter = this.overlays.getIteratorFrom(prefix);
+    while (iter.hasNext()) {
+      const entry = iter.getNext();
+      const overlay = entry.value;
+      const key = overlay.getKey();
+      if (!collection.isPrefixOf(key.path)) {
+        break;
+      }
+      // Documents from sub-collections
+      if (key.path.length !== immediateChildrenPathLength) {
+        continue;
+      }
+      if (overlay.largestBatchId > sinceBatchId) {
+        result.set(overlay.getKey(), overlay);
+      }
+    }
+
+    return PersistencePromise.resolve(result);
+  }
+
+  getOverlaysForCollectionGroup(
+    transaction: PersistenceTransaction,
+    collectionGroup: string,
+    sinceBatchId: number,
+    count: number
+  ): PersistencePromise<Map<DocumentKey, Overlay>> {
+    let batchIdToOverlays: SortedMap<
+      number,
+      Map<DocumentKey, Overlay>
+    > = new SortedMap<number, Map<DocumentKey, Overlay>>(
+      (key1: number, key2: number) => key1 - key2
+    );
+
+    const iter = this.overlays.getIterator();
+    while (iter.hasNext()) {
+      const entry = iter.getNext();
+      const overlay = entry.value;
+      const key = overlay.getKey();
+      if (key.getCollectionGroup() !== collectionGroup) {
```

Comment:
We should consider using the IndexManager to get the list of collection paths first.

Reply:
You can call https://osscs.corp.google.com/firebase-sdk/firebase-js-sdk/+/master:packages/firestore/src/local/index_manager.ts;l=52 and then only iterate the collections that you are interested in. 

This method is not called in practice though so we could punt it. It is only called by the IndexBackfiller, which requires persistence.